evaluator/evaluator.go

```diff
@@ -491,15 +501,37 @@ func applyFunction(fn object.Object, args []object.Object) object.Object {
 		if len(fn.Parameters) != len(args) {
 			return newError("Supplied %v args, but %v are expected", len(args), len(fn.Parameters))
 		}
-		extendedEnv := extendPureFunctionEnv(fn, args)
+		// strip the old metadata off of the incoming params
+		traceableArgs := make([]object.Object, len(fn.Parameters))
+		for i, a := range args {
+			deps := make(map[int]bool)
+			deps[i] = true
+			traceableArgs[i] = a.Copy()
+			traceableArgs[i].SetMetadata(object.TraceMetadata{Dependencies: deps})
+		}
+
+		extendedEnv := extendPureFunctionEnv(fn, traceableArgs)
 		var evaluated object.Object
 		if val, ok := fn.Get(args); ok {
 			evaluated = val
 		} else {
 			evaluated = Eval(fn.Body, extendedEnv)
 			fn.Set(args, evaluated)
 		}
-		return unwrapReturnValue(evaluated)
+		fmt.Printf("deps: %+v\n", evaluated.GetMetadata())
+		res := unwrapReturnValue(evaluated)
+		// now we assign our dependencies for the function call itself
+		// let b = pfn(x, y) { if (x % 2 == 0) { a(x, 2, y) } else { 1 } }
+		fnMetadata := res.GetMetadata()
+		callMetadata := object.TraceMetadata{}
+		for i, a := range args {
+			if obj, ok := fnMetadata.Dependencies[i]; ok && obj == true {
```

Comment:
Do we need the `== true`?

Reply:
So I'm keeping this around in case we want to remove a dependency by setting it's key to false. I figure it can't hurt but would be fine getting rid of it as long as we make a note of the behavior